File: Question 0543 - Diameter of Binary Tree/src/Solution.java
class Solution {
    private int[] diameterAndHeight(TreeNode node) {
        if (node == null) {
            return new int[]{0, 0}; // {diameter, height}
        }

        int[] leftResult = diameterAndHeight(node.left);
        int[] rightResult = diameterAndHeight(node.right);

        int height = 1 + Math.max(leftResult[1], rightResult[1]);
        int diameterThroughRoot = leftResult[1] + rightResult[1] ;
        int diameter = Math.max(Math.max(leftResult[0], rightResult[0]), diameterThroughRoot);

        return new int[]{diameter, height};
    }
    public int diameterOfBinaryTree(TreeNode root) {
        return diameterAndHeight(root)[0];
    }
}

File: Question 0543 - Diameter of Binary Tree/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int solve(TreeNode *root, int &len)
    {
        if (root == NULL)
            return 0;

        int left = solve(root->left, len);
        int right = solve(root->right, len);

        len = max(len, left + right);
        return max(left, right) + 1;
    }

    int diameterOfBinaryTree(TreeNode *root)
    {
        int len = 0;
        solve(root, len);
        return len;
    }
};